spritesheet-manager.js

```diff
@@ -29,13 +33,35 @@ class SpritesheetManager {
     }
     return spritesheet;
   }
+
   async getSpriteSheetForAppUrlAsync(appUrl, opts) {
-    let spritesheet = this.spritesheetCache.get(appUrl);
-    if (!spritesheet) {
-      spritesheet = await this.getSpriteSheetForAppUrlInternal([appUrl, opts]);
-      this.spritesheetCache.set(appUrl, spritesheet);
+    const spritesheet = this.spritesheetCache.get(appUrl);
+
+    if (spritesheet) return spritesheet;
+
+    return new Promise((resolve) => {
+      this.waitQueue.push({ url: appUrl, opts, resolve });
+      this.loadSpriteSheetFromQueue();
+    });
+  }
+
+  loadSpriteSheetFromQueue() {
+    if (this.currentlyLoading >= this.maxConcurrentLoading || this.waitQueue.length <= 0) return;
+
+    const loopCount = Math.min(this.maxConcurrentLoading, this.waitQueue.length);
+    for (let i = 0; i < loopCount; i++) {
+      this.currentlyLoading++;
+      const data = this.waitQueue.splice(0, 1)[0];
+
+      this.getSpriteSheetForAppUrlInternal([ data.url, data.opts ]).then((spritesheet) => {

```

Comment:
For consistency, we aim for `async` wrapping where possible rather than callbacks. I believe VSCode also pushes you to do it.